core/src/main/resources/error/error-classes.json

```
@@ -1537,6 +1537,12 @@
     ],
     "sqlState" : "42883"
   },
+  "UNRESOLVED_USING_COLUMN_FOR_JOIN" : {
+    "message" : [
+      "USING column <colName> cannot be resolved on the <side> side of the join. The <side>-side columns: [<plan>]."
```

USING column `c0` cannot be resolved on the left side of the join. The left-side columns: [c1, c2, c3].

Do we use the suggestion code we have for UNRSOLVED_COLUMN>? We should NOT pump out all columns.
```suggestion
      "USING column <colName> cannot be resolved on the <side> side of the join. The <side>-side columns: [<suggestion>]."
```

---

Sounds good! Will address to use the suggestion.
Thanks!!

---

I updated the code to mimic the behavior of `UNRESOLVED_COLUMN.WITH_SUGGESTION`.
But just realized that `UNRESOLVED_COLUMN.WITH_SUGGESTION` also pump out all columns, so eventually the behavior is not changed.
In anyway the column names should be backquotes, so I did an update.
cc @MaxGekk

---

"But just realized that UNRESOLVED_COLUMN.WITH_SUGGESTION also pump out all column"
Huh? @MaxGekk

---

@itholic Could you restrict the number of columns by sorting the list (to have stable output) and return top 3 columns. Or be more smarter and output 3 columns that starts from the same char as the unresolved column, and if it is empty, just print the first 3 columns of the sorted list.

---

Sounds good.
Can I do this in separate ticket to apply changes for all related places, not only for this PR ?
Just created ticket here: SPARK-42414.

---

> Can I do this in separate ticket to apply changes for all related places, not only for this PR ?

@itholic  Sure.

---

@itholic I think we can do better than pre-fix check. See slack.

---

Oh... we're using `StringUtils.orderStringsBySimilarity` that returns better candidates in `unresolvedColumnError` as below:
```scala
            val orderedCandidates =
              StringUtils.orderStringsBySimilarity(f.name, candidates)
            throw QueryCompilationErrors.unresolvedColumnError(f.name, orderedCandidates)
```

But, we're just pumping out all columns in `unresolvedColumnWithSuggestionError` as below:
```scala
  private def resolveException(colName: String, fields: Array[String]): AnalysisException = {
    QueryCompilationErrors.unresolvedColumnWithSuggestionError(
      colName, fields.map(toSQLId).mkString(", "))
  }
```

Let me integrate `unresolvedColumnWithSuggestionError` into `unresolvedColumnError` so that we can display better suggestions in the error message.

Thanks!

---

Anyway, let me address within SPARK-42414 at once through all code path, after this PR is completed.